README.md
# Garrett

Another conversion tool for audio files (and you probably don't even need it)

![Garrett's user interface](screenshot.png)

## What does this software do?

If you have to ask this question, you're probably not the target audience ;)

Jokes aside. Garrett converts the given audio files into the "Waveform Audio File Format" (WAVE) with a static bit depth of 16 bits.
You can choose between two output sample rates: 44100 Hz and 48000 Hz.
The supported input formats are:
- WAVE
- MP3
- FLAC
- AIFF

## Why the hell is this application existing?

The purpose of this originally was to provide a simple conversion interface without needing a lot of knowledge behind the scenes.
The resulting audio files were primarily meant to be played on the x32 console from Behringer. However, this program can be used for other purposes too.

## Installation

For macOS and Linux, prebuilt binaries are available through [GitHub Releases](https://github.com/hannesbraun/garrett/releases).
It should also work with Windows. However, this is not tested, and you need to build it yourself.

### Building

Make sure to install the required dependencies. This is primarily `libsamplerate`. See the GitHub Actions workflow files for more details.

Install the Fyne utility with

```bash
go get fyne.io/fyne/v2/cmd/fyne
```

Then, package the application (in this example for macOS):

```bash
cd build/macos
bash build.sh 0.1.0
```

Replace `0.1.0` with the version number you're currently building.

## License

Garrett is licensed under the GNU General Public License v3.0. For more information see [LICENSE](LICENSE "GNU General Public License v3.0").

## Author

Hannes Braun (hannesbraun@mail.de)


CHANGELOG.md
# 0.1.0
*Initial version*


converter.go
package main

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"errors"
	"fyne.io/fyne/v2/data/binding"
	"fyne.io/fyne/v2/widget"
	"github.com/dh1tw/gosamplerate"
	"github.com/gabriel-vasile/mimetype"
	"github.com/go-audio/aiff"
	"github.com/go-audio/audio"
	"github.com/hajimehoshi/go-mp3"
	"github.com/mewkiz/flac"
	"github.com/youpy/go-wav"
	"io"
	"math"
	"os"
	"path"
	"strings"
)

// Updates the status label
func updateStatus(statusLabel **widget.Label, text string) {
	if len(text) > 95 {
		// Crop the text to avoid resizing the window
		(*statusLabel).SetText(text[:92] + "...")
	} else {
		(*statusLabel).SetText(text)
	}
}

// Runs the conversions
func convert(files []string, outDir string, sampleRate float64, progress *binding.ExternalFloat, statusLabel **widget.Label) []string {
	failed := make([]string, 0)
	if len(files) <= 0 {
		// No files; nothing to do
		return failed
	}
	progressStep := 1.0 / (3.0 * float64(len(files)))

	for i, file := range files {
		// Update progress
		_ = (*progress).Set(float64(3*i) * progressStep)
		updateStatus(statusLabel, "Decoding "+file)

		// Detect file type
		mimeType, err := mimetype.DetectFile(file)
		if err != nil {
			failed = append(failed, file)
			continue
		}

		// Decode the file
		var track Track
		switch mimeType.String() {
		case "audio/mpeg": // mp3
			track, err = decodeMp3(file)
			if err != nil {
				failed = append(failed, file)
				continue
			}
		case "audio/flac": // flac
			track, err = decodeFlac(file)
			if err != nil {
				failed = append(failed, file)
				continue
			}
		case "audio/wav": // wav
			track, err = decodeWav(file)
			if err != nil {
				failed = append(failed, file)
				continue
			}
		case "audio/aiff": // aiff
			track, err = decodeAiff(file)
			if err != nil {
				failed = append(failed, file)
				continue
			}
		default:
			failed = append(failed, file)
			continue
		}

		_ = (*progress).Set(float64(3*i+1) * progressStep)
		updateStatus(statusLabel, "Resampling "+file)

		// Resample (if necessary)
		resampled := track.data
		if sampleRate != track.sampleRate {
			resampled, err = gosamplerate.Simple(track.data, sampleRate/track.sampleRate, int(track.channels), gosamplerate.SRC_SINC_BEST_QUALITY)
			if err != nil {
				failed = append(failed, file)
				continue
			}
		}

		_ = (*progress).Set(float64(3*i+2) * progressStep)
		updateStatus(statusLabel, "Assembling wave samples")

		// Assemble wav samples
		var samples []wav.Sample
		for i := 0; i < len(resampled); i += 2 {
			var sample [2]int
			// Clipping is necessary since the signal may leave the default range from -1.0 to 1.0 while resampling
			sample[0] = clip16Bit(int(resampled[i] * math.MaxInt16))
			if i+1 < len(resampled) {
				sample[1] = clip16Bit(int(resampled[i+1] * math.MaxInt16))
			} else {
				sample[1] = 0
			}

			samples = append(samples, wav.Sample{Values: sample})
		}

		// Figure out the destination path
		baseName := path.Base(file)
		indexSuffix := strings.LastIndex(baseName, ".")
		if indexSuffix > 0 {
			baseName = baseName[0:indexSuffix]
		}

		updateStatus(statusLabel, "Writing "+path.Join(outDir, baseName+".wav"))

		// Write wav file
		out, err := os.Create(path.Join(outDir, baseName+".wav"))
		if err != nil {
			failed = append(failed, file)
			continue
		}
		wavOut := bufio.NewWriter(out)
		writer := wav.NewWriter(wavOut, uint32(len(samples)), track.channels, uint32(sampleRate), 16)
		err = writer.WriteSamples(samples)
		_ = out.Close()
		if err != nil {
			failed = append(failed, file)
			continue
		}

		_ = out.Close()
	}

	_ = (*progress).Set(1.0)
	updateStatus(statusLabel, "Idle")

	return failed
}

func clip16Bit(sample int) int {
	if sample < -32768 {
		return -32768
	} else if sample > 32767 {
		return 32767
	} else {
		return sample
	}
}

type Track struct {
	data       []float32
	sampleRate float64
	channels   uint16
}

func decodeMp3(file string) (Track, error) {
	f, err := os.Open(file)
	if err != nil {
		return Track{}, err
	}
	defer f.Close()

	d, err := mp3.NewDecoder(f)
	if err != nil {
		return Track{}, err
	}

	buf := make([]byte, d.Length())
	i := int64(0)
	for {
		read, err := d.Read(buf[i:])
		if err == io.EOF {
			break
		} else if err != nil {
			return Track{}, err
		}
		i += int64(read)
	}

	floatBuf := make([]float32, d.Length()/2)
	for i = 0; i < d.Length(); i += 2 {
		var sample int16
		_ = binary.Read(bytes.NewBuffer(buf[i:i+2]), binary.LittleEndian, &sample)
		floatBuf[i/2] = float32(sample) / math.MaxInt16
	}

	sampleRate := float64(d.SampleRate())

	return Track{
		data:       floatBuf,
		sampleRate: sampleRate,
		channels:   2,
	}, nil
}

func decodeFlac(file string) (Track, error) {
	stream, err := flac.Open(file)
	if err != nil {
		return Track{}, err
	}
	defer stream.Close()

	sampleRate := uint32(44100)
	dualChannel := false
	buf := make([]float32, 0)
	for {
		frame, err := stream.ParseNext()
		if err == io.EOF {
			break
		} else if err != nil {
			return Track{}, err
		}

		frameBuf := make([]float32, 0)
		maxIntVal := 1 << (frame.BitsPerSample - 1)
		for i := 0; i < frame.Subframes[0].NSamples; i++ {
			frameBuf = append(frameBuf, float32(frame.Subframes[0].Samples[i])/float32(maxIntVal))
			if len(frame.Subframes) > 1 {
				frameBuf = append(frameBuf, float32(frame.Subframes[1].Samples[i])/float32(maxIntVal))
				dualChannel = true
			}
		}
		buf = append(buf, frameBuf...)
		sampleRate = frame.SampleRate
	}

	channels := uint16(2)
	if !dualChannel {
		channels = 1
	}

	return Track{
		data:       buf,
		sampleRate: float64(sampleRate),
		channels:   channels,
	}, nil
}

func decodeWav(file string) (Track, error) {
	f, err := os.Open(file)
	if err != nil {
		return Track{}, err
	}
	defer f.Close()

	reader := wav.NewReader(f)
	format, err := reader.Format()
	if err != nil {
		return Track{}, err
	}
	channels := format.NumChannels

	floatBuf := make([]float32, 0)
	for {
		// Reading samples
		samples, err := reader.ReadSamples()
		if err == io.EOF {
			break
		}

		// Converting to float32
		for _, sample := range samples {
			// Iterate over channels
			for i := uint(0); i < 2; i++ {
				floatBuf = append(floatBuf, float32(reader.FloatValue(sample, i)))
			}
		}
	}

	outChannels := uint16(2)
	if channels == 1 {
		outChannels = 1
	}

	return Track{
		data:       floatBuf,
		sampleRate: float64(format.SampleRate),
		channels:   outChannels,
	}, nil
}

func decodeAiff(file string) (Track, error) {
	f, err := os.Open(file)
	if err != nil {
		return Track{}, err
	}
	defer f.Close()

	decoder := aiff.NewDecoder(io.ReadSeeker(f))
	if !decoder.IsValidFile() {
		return Track{}, errors.New("invalid aiff file")
	}

	floatBuf := make([]float32, 0)
	intBuf := make([]int, 255)
	ch := 0
	buf := &audio.IntBuffer{Data: intBuf}
	for {
		n, err := decoder.PCMBuffer(buf)
		if n == 0 {
			break
		}
		if err != nil {
			return Track{}, err
		}

		for _, val := range buf.AsFloat32Buffer().Data {
			if ch < 2 {
				floatBuf = append(floatBuf, val)
			}
			ch = (ch + 1) % int(decoder.NumChans)
		}
	}

	outChannels := uint16(2)
	if decoder.NumChans == 1 {
		outChannels = 1
	}

	return Track{
		data:       floatBuf,
		sampleRate: float64(decoder.SampleRate),
		channels:   outChannels,
	}, nil
}


selectFiles.go
package main

import (
	"github.com/gabriel-vasile/mimetype"
	"io/ioutil"
	"path"
)

var supportedMimeTypes = []string{
	"audio/mpeg",
	"audio/flac",
	"audio/wav",
	"audio/aiff",
}

// Get convertable files from a directory (recursively)
func filesFromDirectory(dir string) []string {
	result := make([]string, 0)

	items, err := ioutil.ReadDir(dir)
	if err != nil {
		return result
	}

	for _, item := range items {
		if !item.IsDir() {
			mimeType, err := mimetype.DetectFile(path.Join(dir, item.Name()))
			if err != nil {
				continue
			}

			if isSupportedMimeType(mimeType) {
				result = append(result, path.Join(dir, item.Name()))
			}
		} else {
			// Search subdirectory
			result = append(result, filesFromDirectory(path.Join(dir, item.Name()))...)
		}
	}

	return result
}

func isSupportedMimeType(mimeType *mimetype.MIME) bool {
	for _, supportedMimeType := range supportedMimeTypes {
		if supportedMimeType == mimeType.String() {
			return true
		}
	}

	return false
}


main.go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/canvas"
	"fyne.io/fyne/v2/container"
	"fyne.io/fyne/v2/data/binding"
	"fyne.io/fyne/v2/dialog"
	"fyne.io/fyne/v2/layout"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
	"github.com/gabriel-vasile/mimetype"
	"image/color"
	"net/url"
	"os"
	"path"
	"strconv"
	"strings"
)

var VERSION = "0.1.0"

// Custom theming

var primaryColor = color.RGBA{
	R: 0,
	G: 0,
	B: 0,
	A: 0,
}

type garrettTheme struct{}

func (m garrettTheme) Color(name fyne.ThemeColorName, variant fyne.ThemeVariant) color.Color {
	if name == theme.ColorNamePrimary && primaryColor.A != 0 {
		return primaryColor
	}

	return theme.DefaultTheme().Color(name, variant)
}

func (m garrettTheme) Icon(name fyne.ThemeIconName) fyne.Resource {
	return theme.DefaultTheme().Icon(name)
}

func (garrettTheme) Font(style fyne.TextStyle) fyne.Resource {
	return theme.DefaultTheme().Font(style)
}

func (garrettTheme) Size(name fyne.ThemeSizeName) float32 {
	switch name {
	case theme.SizeNameText:
		return 12
	case theme.SizeNameInlineIcon:
		return 16
	default:
		return theme.DefaultTheme().Size(name)
	}
}

func main() {
	if len(os.Args) > 1 {
		value, err := strconv.ParseInt(strings.TrimPrefix(os.Args[1], "0x"), 16, 64)
		if err == nil {
			primaryColor = color.RGBA{
				R: uint8((0xFF0000 & value) >> 16),
				G: uint8((0x00FF00 & value) >> 8),
				B: uint8(0x0000FF & value),
				A: 255,
			}
		}
	}

	garrettApp := app.New()
	garrettApp.Settings().SetTheme(&garrettTheme{})

	w := garrettApp.NewWindow("Garrett")

	// File list
	inputFiles := binding.BindStringList(
		&[]string{},
	)
	selected := 0
	list := widget.NewListWithData(inputFiles,
		func() fyne.CanvasObject {
			return widget.NewLabel("template")
		},
		func(i binding.DataItem, o fyne.CanvasObject) {
			o.(*widget.Label).Bind(i.(binding.String))
		},
	)
	list.OnSelected = func(id widget.ListItemID) {
		selected = id
	}
	list.OnUnselected = func(id widget.ListItemID) {
		selected = inputFiles.Length() - 1
	}

	// List control buttons
	addButton := widget.NewButton("Add...", func() {
		dialog.ShowFileOpen(func(closer fyne.URIReadCloser, err error) {
			if closer == nil || err != nil {
				return
			}

			mimeType, err := mimetype.DetectFile(closer.URI().Path())
			if isSupportedMimeType(mimeType) && err == nil {
				err := inputFiles.Append(closer.URI().Path())
				if err != nil {
					dialog.ShowInformation("Error", "Unable to append the file "+path.Base(closer.URI().Path())+" to the list", w)
				}
			} else {
				mimeTypeStr := "<unknown>"
				if mimeType != nil {
					mimeTypeStr = mimeType.String()
				}
				dialog.ShowInformation("Not supported", "Converting the file "+path.Base(closer.URI().Path())+" is not supported.\nUnsupported type: "+mimeTypeStr, w)
			}
		}, w)
	})
	addDirButton := widget.NewButton("Add directory...", func() {
		dialog.ShowFolderOpen(func(uri fyne.ListableURI, err error) {
			if uri == nil || err != nil {
				return
			}

			items := filesFromDirectory(uri.Path())
			errOccurred := false
			for _, item := range items {
				err := inputFiles.Append(item)
				if err != nil {
					errOccurred = true
				}
			}

			if errOccurred {
				dialog.ShowInformation("Error", "Unable to append some files", w)
			}
		}, w)
	})
	removeButton := widget.NewButton("Remove", func() {
		if selected >= inputFiles.Length() {
			return
		}

		in, _ := inputFiles.Get()
		in = append(in[:selected], in[selected+1:]...)
		err := inputFiles.Set(in)
		if err != nil {
			dialog.ShowInformation("Error", "Unable to remove the file "+path.Base(in[selected]), w)
			return
		}

		if selected >= inputFiles.Length() {
			selected = inputFiles.Length() - 1
		}
	})
	clearButton := widget.NewButton("Clear", func() {
		in, _ := inputFiles.Get()
		err := inputFiles.Set(in[:0])
		if err != nil {
			dialog.ShowInformation("Error", "Unable to remove the file "+path.Base(in[selected]), w)
			return
		}
		selected = 0
	})
	listControls := container.NewHBox(addButton, addDirButton, removeButton, clearButton)

	// Output
	outDirLabel := widget.NewLabel("Output directory")
	outDir, err := os.UserHomeDir()
	if err != nil {
		outDir = "/"
	}
	outDirLabel2 := widget.NewLabel(outDir)
	selectOutDirButton := widget.NewButtonWithIcon("", theme.FolderIcon(), func() {
		dialog.ShowFolderOpen(func(uri fyne.ListableURI, err error) {
			if uri == nil || err != nil {
				return
			}

			outDir = uri.Path()
			outDirLabel2.SetText(outDir)
		}, w)
	})
	outDirVal := container.NewBorder(nil, nil, nil, selectOutDirButton, outDirLabel2)

	// Sample rate selection
	sampleRateLabel := widget.NewLabel("Sample rate")
	sampleRate := 48000
	sampleRateButtons := widget.NewRadioGroup([]string{"44100 Hz", "48000 Hz"}, func(val string) {
		if val == "44100 Hz" {
			sampleRate = 44100
		} else {
			sampleRate = 48000
		}
	})
	sampleRateButtons.SetSelected("48000 Hz")

	// Assemble form
	form := container.New(layout.NewFormLayout(), outDirLabel, outDirVal, sampleRateLabel, sampleRateButtons)

	// Status and control "bar"
	progressVal := 0.0
	progress := binding.BindFloat(&progressVal)
	progressBar := widget.NewProgressBarWithData(progress)
	statusLabel := widget.NewLabel("Idle")
	running := false
	startButton := widget.NewButton("Convert", func() {
		if running {
			return
		}
		running = true

		files, _ := inputFiles.Get()
		go func(files []string, outDir string, sampleRate int, progress *binding.ExternalFloat, statusLabel **widget.Label) {
			failed := convert(files, outDir, float64(sampleRate), progress, statusLabel)
			if len(failed) > 0 {
				filesStr := strings.Join(failed, "\n")
				dialog.ShowInformation("Unable to convert the following files", filesStr, w)
			}
			running = false
		}(files, outDir, sampleRate, &progress, &statusLabel)
	})
	bottomContainer := container.NewBorder(progressBar, nil, nil, startButton, statusLabel)

	// "About"
	copyrightLabel := canvas.NewText("  Copyright (C) 2021 Hannes Braun | Version "+VERSION, garrettApp.Settings().Theme().Color(theme.ColorNameForeground, garrettApp.Settings().ThemeVariant()))
	repoUrl, _ := url.Parse("https://github.com/hannesbraun/garrett")
	repoHyperlink := widget.NewHyperlink("Visit this application's git repository", repoUrl)

	// Put everything together and run it
	content := container.NewBorder(nil, container.NewVBox(listControls, form, dummy(), widget.NewSeparator(), dummy(), bottomContainer, dummy(), widget.NewSeparator(), dummy(), copyrightLabel, repoHyperlink), nil, nil, list)
	w.SetContent(content)
	w.Resize(fyne.NewSize(768, 700))
	w.ShowAndRun()
}

func dummy() fyne.CanvasObject {
	dummy := canvas.NewRectangle(color.RGBA{A: 0})
	dummy.Resize(fyne.Size{
		Width:  0,
		Height: 0,
	})
	dummy.Show()
	return dummy
}
